### PR TITLE
Layerwise constraints and regularization

### PR DESCRIPTION
**Pull request by the-moliver**
I've moved the functions to relevant modules which can be used as below. And I've separated the weight and bias regularizers for better clarity, with the defaults being no regularization or constraints.
I don't think it makes sense to have the regularizers and constraints specified by strings, i.e. W_constraint='maxnorm' , since the user will usually want to specify the parameter and it's not clear how to do that with the string convention. It works fine when the defaults are probably fine, as with the initializations. But even in that case, how would one specify the scale for the initialization ```init='normal' ``` if you wanted to change it? The way I have it below seems clearer and easier to work with and is similar to the way the dropout rate is specified. Let me know what you think and if you think I should change something.

```python
from keras.constraints import maxnorm
from keras.regularizers import l2

model = Sequential()
model.add(Dense(784, 1000, W_regularizer=l2(.05), W_constraint=maxnorm(2)))

### CODE REVIEW

**Comment by fchollet** [comment]
The main issue I have with this implementation is that it would break every layer not updated here (so every layer outside of layers.core). It would also add quite a bit of overhead/boilerplate to the creation of new layers, since you would have to define regularizer and constraint attributes every time (and for many layers, these are not even relevant). We want layer creation to be as simple and straightforward as possible, to allow for easy extensibility.

A good solution would be to make constraint and regularization completely optional, and only include them in the Dense layer. Features should be modular, and that also means that a module that does not use a certain feature should *not* have to care about it (or even know about it). So we need to find a way to make this new feature truly modular.

Another thing I'm noticing is that you are now modifying every single optimizer with the same 3 lines of code. So these 3 lines probably belong in the parent class, not in every descendant. Like layers, optimizers should be easy to build, and free of boilerplate (as much as possible).

**Comment by the-moliver** [comment]
My counter argument is that in (almost) any layer where params are defined, it makes sense to have the ability to regularize and or constrain them. There are many layers with no params which consequently include the boilerplate ```self.params = []```, so the addition of two additional and related lines of boilerplate seemed most consistent with what already existed. But, you are right I should add it to the rest of the layers so they aren't broken, and give the ability to regularize in the other layers as well where appropriate (such as convolution and rnn). I don't know that there is a way to cleanly make them optional and otherwise invisible, just as the params are not optional and must be defined as an empty list when not needed. But if you have a better solution, I would argue that this solution should be applied to params as well. Let me know, I'm very curious if there's a better way to do all of this!

As far as the optimizers, I'm not sure what should go in the parent class. The first line I modified 
```p, g, r, c in zip(params, grads, regularizers, constraints)``` 
I just added ```regularizers``` and ```constraints```, which are lists of functions, to the lists of params and grads. The next line ```g = r(g,p)``` is just applying the regularizer function to the gradient. I could potentially move this to the get_gradients function and loop over zip(params, grads, regularizers) so the get_gradients function returns the regularized gradients, somewhat similarly to what you have now. But the application of the constraint function, ```new_p = c(new_p)```, needs to be applied before the updates are created by ```updates.append((p, new_p))``` and so must be done in the optimizer loop. Thus it seemed cleaner to do the application of the regularizer and the constraints in a similar way. This also separates the get_gradients function from regularization, which seemed to me conceptually clearer.

I really do appreciate your desire to keep things as simple and clean as possible and have tried to add what I consider important functionality with this in mind. But of course let me know if you think there's a better way.

**Comment by fchollet** [comment]
It's a purely economic issue. Adding anything to the codebase has benefits (the functionality it brings), and it has a cost: it might hinder usability, it might complicate future development, it creates maintenance liability, etc. We make the addition only when it is clear that the benefits outweigh the costs.

The benefits are multiplied by how many people would use the functionality, and how much value it would bring them. The cost is multiplied by how far-reaching the architecture & development implications are. 

Something that affects (and complicates) every layer and every optimizer, not just current ones, but all that will ever be developed, has a very high cost. This is especially true for Keras, since simplicity and minimalism are distinctive features of Keras, its competitive advantage. *Decreasing the simplicity of the interface and architecture in any way is damaging the long term chances of success of Keras.*

Meanwhile the benefits do not seem very high, as regularization in deep nets is pretty much always done with Dropout, not L1/L2 reg. Besides the existing optimizers already support a (non-layer specific) way to use L1/L2 reg. Adding useful customizability to a rarely used feature is definitely valuable, but not a priority. 

> There are many layers with no params which consequently include the boilerplate self.params = [], so the addition of two additional and related lines of boilerplate seemed most consistent with what already existed. 

One line of boilerplate is not an issue, because it would take at the very least one line to abstract it away, and I'd rather avoid unnecessary abstraction. If we have more than one line, then we should abstract it, which is easy: put all the boilerplate in the \__init\__ of the parent class and call super().\__init\__ in each layer. Back to a single line of boilerplate (a less friendly one, though). 

It is quite possible that we'll end up doing this. 

Regarding the optimizers: currently constraints and regularizations are handled by the parent class is a transparent way, so it's easy to build new optimizers. We'd like to keep it that way. I will think of a way to handle custom per-parameter constraints/regularizations in the same fashion.

**Comment by the-moliver** [comment]
I've updated the PR based on your suggestions and concerns. I agree that L1/L2 penalties are not as important these day, but maxnorm regularization is quite important, and the current implementation in keras doesn't seem to work. I started this to fix that issue and add related features. Also for something I'm working on I needed a non-negative constraint at a particular layer. In this PR the new features require no additional lines of boilerplate than currently when creating new layers/optimizers, just a little bit different lines of code.

I also fixed an issue where keyword arguments to optimizers weren't being used, i.e. the clipnorm argument  in ```adam = Adam(clipnorm=2)``` wasn't being registered by the function because the line ```self.__dict__.update(kwargs)``` was missing.

**Comment by fchollet** [comment]
This is looking very good. A few comments...

In much the same way that regularization is handled at the beginning of the optimizer call in one line, we can handle constraints in one line at the end, without touching the update construction loop. Such as:
```python
updates = self.apply_constraints(updates, constraints)
return updates
```
Where apply_constraints would amount to:
```python
return [(p, c(new_p)) for (p, new_p), c in zip(updates, constraints)]
```
Once we have that, the only impact on optimizers will just be two very reasonable lines. Which means optimizers will stay developer-friendly : )

Otherwise, calling super in every layer is quite reasonable. We were probably going to need this sooner or later anyway, as we add default layer functionality. 

Esthetically, "identity" seems like it would be more in line with Keras naming conventions than "ident". 

I'm curious as to whether there is any efficiency impact in applying identity in the (most frequent) cases where no regularization/constraint is applied. I would expect Theano to optimize it away, but it would be useful to compare speed before and after this PR.

And finally, two important things before we are able to merge these changes: 1) tests and 2) documentation. As you will have noticed with the clipnorm issue in Adam, lack of tests is a massive issue in ensuring code quality, and we're aiming at full unit test coverage before v1.0. Since the changes in this PR are pretty consequent, it would be great to have a way to verify that they work as intended and are bug-free. We are going to write tests for all features sooner or later, so we might as well start writing them for any new functionality being added now.

**Comment by the-moliver** [comment]
Thanks for the comments! 
I changed ident to identity as requested. As far as I could tell the overhead from the identity function was measured in nanoseconds, so basically just the same as an assignment.

I thought about implementing the constraints as you described, but it was somewhat problematic as  ```updates``` doesn't just contain parameter updates, but also various accumulators depending on the algorithm. Matching the constraints up to the parameter updates within ```updates``` would be tricky and ugly I think. The parameter updates could be kept separate, in a variable ```param_updates``` until the constraints are applied as you suggest and then appended to ```updates```, but that would be more lines of code than now. I thought the cleanest solution was the update_params function I created. Let me know if you can think of a better way.

In this updated PR, I also added a generalized KL divergence penalty to add the functionality requested in https://github.com/fchollet/keras/issues/87 . This was surprisingly easy, requiring a few minor changes. Thank god for Theano's autodiff. It can be used as below. ```target``` is the desired average activation, and ```beta``` is the weighting used on this penalty. Both are optional and turned off by default. I'm not in love with the name ```beta``` so let me know if you'd rather it be called something else.

```python
# use relu activation with a desired average activation of 0.2
# beta is set to 0.1 by default
model.add(Activation('relu', target=.2))

Finally, I totally agree about tests and documentation. Can you send me an example of the model I should follow for each?

**Comment by fchollet** [comment]
> Finally, I totally agree about tests and documentation. Can you send me an example of the model I should follow for each?

For documentation, you would just need to update the [optimizers.md](https://github.com/fchollet/keras/blob/master/docs/sources/optimizers.md) page and the [layers](https://github.com/fchollet/keras/tree/master/docs/sources/layers) documentation. If you feel this is necessary, you may need to create new pages for constraints and regularizers.

For tests, we don't have a template yet (if we did, if would be very helpful...). So you can just create a new file in /tests/ with your test functions. Only requirements is that the tests should be fast, and should allow to check that regularization and constraints work as expected.

**Comment by fchollet** [comment]
I only just noticed you added the docs and tests... I'll review everything and get this merged asap.

**Comment by the-moliver** [comment]
I'm still working on it a bit, I'll let you know tomorrow when it's done.

**Comment by pranv** [comment]
Great Job Man!

**Comment by the-moliver** [comment]
Ok, everything is being tested now! Let me know!

**Comment by fchollet** [comment]
Merged : )

I made a couple changes; the data download process for MNIST was breaking Python3 compatibility. I also made the MNIST example reproducible, with a test accuracy of 0.983 (same training time), which is quite competitive for a simple MLP.

**Comment by fchollet** [comment]
Commit af46ab0d4dc981b899fe6c82c1712e95f251a118 appears to have broken the Kaggle Otto example, most likely at optimizer level. Since that example isn't doing anything special, I assume a wide range of models would be broken as well.

Try it: checkout at the previous commit, e7da65f1d65fe363c945792d5ba978d5cc6c6a86 -> learning happens normally. 
Checkout out at af46ab0d4dc981b899fe6c82c1712e95f251a118 -> learning is broken. Not amount of tweaking the example (changing optimizers, removing BatchNorm/PReLU/etc) appears to make it work. The issue is almost certainly a broken optimizer algo.

Can you figure this out? If we can't fix it quickly I'll revert all optimizers to their last known working state.

**Comment by fchollet** [comment]
Reverted (models.py and optimizers.py only) until we figure out a fix. 

As far as I can tell, the issue is that empty lists for constraints are messing up the zip() happening in the main loop of every optimizer. I'll look at it in more detail later.

**Comment by the-moliver** [comment]
Can you tell me what you mean by broken? The example seems to run just fine for me.

**Comment by fchollet** [comment]
The loss is stuck at ~2.10, which is what you would get from a random input. 

It appears this can be fixed for a given optimizer by replacing:
```python
for p, g, a, c in zip(params, grads, accumulators, constraints):
```
and 
```python
self.update_params(p, new_p, updates, c)
```
with 
```python
for p, g, a in zip(params, grads, accumulators):
```
and
```python
updates.append((p, new_p))
```

An empty list of constraints appears to be creating the problem.

**Comment by the-moliver** [comment]
Ah I figured it out. Basically models.py needs to add identity to the list of regularizers and constraints if that field is empty in a layer (as by default). That should be the only change needed. The problem was the list of params did not match the list of regularizers and constraints.

**Comment by fchollet** [comment]
Yes, seemed like it. I will fix this when I get a bit of time. I wonder if there would be a way to skip constraints and regularization without necessarily having to apply identity every time...

**Comment by fchollet** [comment]
I'm implementing the following way of configuring layerwise constraints (resp. regularizers), to remove any ambiguity:

- layer.constraint -> the constraint will be applied to all parameters in the layer
- list layer.constraints -> specify per-parameter constraints. The list must have the same length as layer.params. Any "None" in the list will be replaced with identity.
- nothing -> all identity

**Comment by fchollet** [comment]
The changes are back on master, all cleaned up. I got rid of KL divergence because I wasn't happy with the implementation (shipping the layers to the optimizer just didn't feel right). We may re-introduce it later in a simpler form.
